<a href="https://colab.research.google.com/github/adinsa1/Data110/blob/main/201BayesianApproach__2_(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Reverend Thomas Bayes**

<img src="https://upload.wikimedia.org/wikipedia/commons/d/d4/Thomas_Bayes.gif">

## Bayes Theorem

The bayes theorem can be easily derived from conditional probability and is shown as the last formula below.

<img src="https://sites.google.com/site/artificialcortext/_/rsrc/1270096677974/others/mathematics/bayes-theorem/bayes-der.jpg">

### Example

Let us roll a fair die:

B = event of getting an odd number

A= event of getting 1

P(A|B) = P(B|A)*P(A) / P(B)

P(B|A) = 1, since the probability of getting an odd number is certain, given that we got 1

P(A) = 1/6 and P(B) = 1/2 (half of the numbers on the die are odd). Plugging in the numbers into Bayes' formula gives us: P(A|B) = 1/3



<img src = "https://i.pinimg.com/originals/5d/90/79/5d9079798db012ca2af8806d7d211fdf.png" >

## Fequentist

There are different definitions of what we mean by a probability.

Frequentists define probability as the proportion of times (frequency) an event occurs if the trials are repeated many times. This can apply to tossing a coin. But there are events that we cannot try many times, e.g., what is the probability that humans caused global warming?

A population parameter, e.g., the average height of the population is fixed although we don't know it. There is no uncertainty or probability of the parameter. Similarly, a hypothesis is either true or false; no probability of a hypothesis being true.

Frequentists use the maximum of the likelihoods P(D|H) to choose a hypothesis. What you have learned so far is frequentist statistics (classical statistics):
P-values, confidence intervals, Null hypothesis testing.

## Bayesian
Bayesians define probabilty as a quantification of a personal belief. They assign a probability to a hypothesis or to a population parameter or to events that are not repeatable or events that happened in the past.

The probability of a hypothesis is updated as soon as new evidence (data) comes in

<img src="https://1.bp.blogspot.com/-bjvuMHBr_60/XVQXnNfw7NI/AAAAAAAAPzw/TCQGzgDMoNQka7KZAVbMjLQMhmFdza4zgCLcBGAs/w1200-h630-p-k-no-nu/Screenshot%2B2019-08-14%2B10.15.52.png">

## Example

A medical lab test results positive for 99% of all sick people,i.e., P(test positive | sick person) = 0.99

The lab test results negative for 98% of all healthy people, i.e., P(test negative | healthy person) = 0.98

The disease prevalence in the population is 1 in 1000 persons, i.e., P(sick person) = 0.001

If someone gets a positive result, what is the propability that the person has the disease?

Many docotrs overestimate this probability:

https://jamanetwork.com/journals/jamainternalmedicine/fullarticle/1861033

https://www.washingtonpost.com/news/posteverything/wp/2018/10/05/feature/doctors-are-surprisingly-bad-at-reading-lab-results-its-putting-us-all-at-risk/?noredirect=on&utm_term=.51f87fcf8e25


Let us use Bayes formula to answer the question:

P(person sick | test positive) = P(test positive| person sick)*P(person sick)/ p(test positive)

Let us work on the denominator:

P(test positive) = P(test positive | person sick) * P(person sick) + P(test positive | person healthy)*P(person healthy)

Putting together:

P(person sick | test positive) = (0.99 * 0.001) / (0.99*0.001 + 0.02*0.999) = 0.047 or 4.7%



### Three competing hypotheses:

A friend has three coins: The first coin has heads and tails with equal probability, the second one has two heads and the third one has two tails.  He chose one of the coins randomly and tossed it. It landed heads up. What is the probability that he chose the first or the second or the third coin?

We have competing hypotheses (explanations) as to which coin he chose.

### Prior P(H)

In [ ]:
import numpy as np
# 0 ->fair coin HT
# 1-> two heads coin HH
# 2 -> two tils coin TT
prior = np.array([1/3,1/3,1/3])




### Likelihood P(D|H)


In [ ]:
# Likelihood for heads: P(heads | first coin) = 1/2; P(heads | second coin) = 1; P(heads|third coin) = 0
likelihood_h = np.array([1/2,1,0])

# Likelihood for tails
likelihood_t = np.array([1/2,0,1])


### Posterior P(H|D) = likelihood * Prior / P(D)

In [ ]:
# We observe heads H:
# We update our probability for each hypothesis in light of the new evidence. Change your mind !
posterior = list([1/3,1/3,1/3])
for i in range(3):
    posterior[i] = likelihood_h[i] * prior[i]  / np.sum(likelihood_h * prior)
print(posterior)

[0.3333333333333333, 0.6666666666666666, 0.0]


The probability of the first coin didn't change. The probability of the second coin doubled. The probability of the third coin changed to zero. If we observe heads, it can be the two tails coin.

In [ ]:
# We observe again heads H. The previous posterior is the new prior. Again we change our minds!
prior = posterior
for i in range(3):
    posterior[i] = likelihood_h[i] * prior[i]  / np.sum(likelihood_h * prior)
print(posterior)

[0.2, 0.8695652173913043, 0.0]


The probability of the first coin decreases, that of the second coin increases, and third coin stays zero.

In [ ]:
# We observe again heads H. The previous posterior is the new prior
prior = posterior
for i in range(3):
    posterior[i] = likelihood_h[i] * prior[i]  / np.sum(likelihood_h * prior)
print(posterior)

[0.1031390134529148, 0.9440152397079056, 0.0]


Now it looks like the first coin is improbable (p=0.103) and the second coin is almost certain (p=0.944). It is certainly not the third coin (p=0).

What happens if the friend tosses the coin one more time and it lands tails up?

Here is the wisdom of the detective Sherlock Holmes:

<img src="https://i.pinimg.com/originals/8b/12/29/8b122907f6ddafbf1aea70d7ee6076e0.jpg">

In [ ]:
# We observe tails T. The previous posterior is the new prior.
prior = posterior
for i in range(3):
    posterior[i] = likelihood_t[i] * prior[i]  / np.sum(likelihood_t * prior)
print(posterior)

[1.0, 0.0, 0.0]


The improbable first coin with heads and tails must be the truth (p=1), since the others were eliminated as impossible (p=0)

### Monty Hall

<img src="https://paulvanderlaken.files.wordpress.com/2020/04/1_znfmn84nnlhyx5tfdblzg.jpeg">

Suppose you're on a game show, and you're given the choice of three doors: Behind one door is a car; behind the others, goats. You pick a door, say No. 1, and the host, who knows what's behind the doors, opens another door, say No. 3, which has a goat. He then says to you, "Do you want to pick door No. 2?" Is it to your advantage to switch your choice?

https://en.wikipedia.org/wiki/Monty_Hall_problem


#### Counterintuitive

Marilyn vos Savant gave the correct answer in her magazine column. Many PhDs "chastised her for misleading and corrupting the
public". Some called her an idiot. But she was right !

The renowned mathematician Paul Erdős was sceptical of the answer. He was reluctantly convinced with a computer simulation.

https://sites.oxy.edu/lengyel/M372/Vazsonyi2003/vazs30_1.pdf


### Answer

Yes, it is to your advantage to switch your choice.

### Simulation (repeated trials of a frequentist)

In [ ]:
## Simulation

import numpy as np

no_change_strategy_wins = 0
change_strategy_wins = 0

all_doors = [1,2,3]

for i in range(10000):
  first_time_pick = np.random.choice(all_doors)
  door_with_car = np.random.choice(all_doors)
  all_doors_n = all_doors.copy()
  if first_time_pick == door_with_car:
    x = all_doors_n.remove(first_time_pick)
    monty_opens_door = np.random.choice(all_doors_n)
  else:
    remain_doors = all_doors_n.remove(first_time_pick)
    w = all_doors_n.remove(door_with_car)
    monty_opens_door =  np.random.choice(all_doors_n)
  second_time_pick = first_time_pick
  if (second_time_pick == door_with_car):
    no_change_strategy_wins = no_change_strategy_wins + 1
  all_doors_n1 = all_doors.copy()
  # Always change strategy
  some_doors = all_doors_n1.remove(first_time_pick)
  y = all_doors_n1.remove(monty_opens_door)
  second_time_pick = all_doors_n1
  if (second_time_pick == door_with_car):
     change_strategy_wins=change_strategy_wins + 1



In [ ]:
print(change_strategy_wins/10000)
print(no_change_strategy_wins/10000)

0.6608
0.3392


It is to your advantage to switch your choice. The probability of winning is 2/3  if you switch and 1/3 if you stick to your first choice.

### Bayesian Solution

Let us assume (without loss of generality) that you pick door 1 and the host Monty opens door 2.

Let A be the event that the car is in door 1

Let B be the event that the car is in door 2

Let C be the event that the car is in door 3

Let D be the event that Monty opens door 2

According to Bayes, the probabilty of car in door 1 given Monty opens door 2 is:
P(A|D) = P(D|A)*P(A) / P(D)

Let us work on the denominator P(D) = P(D|A)*P(A) + P(D|B)P(B) + P(D|C)P(C)

We got the algebra of Bayes rule, but what do these terms stand for.

Let us begin with the numerator: P(D|A) means the probability that Monty opens door 2 given the car is in door 1. This is 1/2, since Monty can choose either of the remaining doors with the goats. Also, P(A) = 1/3, since the car can be in any door equally likely. So the numerator is 1/2 * 1/3 = 1/6

Now the terms of the denominator:

P(D|A)*P(A) is same as the numerator; it is 1/6

P(D|B)*P(B) : First P(D|B) means the probability that Monty opens door 2 given the car is in door 2. That is impossible. Why would he show where the car is?
So P(D|B) = 0 and P(D|B)*P(B) = 0*1/3 = 0

P(D|C)P(C): First P(D|C) means the probability that Monty opens door 2 given the car is in door 3. That is certain, since he cannot open the door with the car (3). He is forced to open door 2 with a goat. So P(D|B ) = 1 and P(D|C)P(C) = 1*1/3 = 1/3.

Now we have the values of all terms, so P(A|D) = 1/6  / (1/6 + 0 + 1/3) = 1/3

A similar calculation shows that P(C|D) = 2/3 or you can see that P(C|D) = 1 - P(A|D) = 1 - 1/3 = 2/3

Switching has P = 2/3 and sticking to first choice has p = 1/3



If you like the Monty Hall problem, here are similar ones with Bayesian solutions:
http://allendowney.blogspot.com/2011/10/all-your-bayes-are-belong-to-us.html


### Bayesian Machine Learning

Let us consider the case of simple linear regression. We have values of X and Y (can be displayed on a scatterplot). We fit a linear regression: y = m*x + b

Remember the slope m and intercept b are statistic of a sample. Classical statistics (frequentist) wants to infer the population parameters for slope and intercept. The Null hypothesis says that nothing is going on, i.e., slope = 0.
We calculate a P-value and if it is less than a certain threshold, we reject the Null hypothesis.

Bayesians agree with the frequentists that there is a fixed (but unknown) population parameter (slope) but they express their uncertainty in a probability distribution. They start with a prior probability distribution of the parameter and adjust it as soon as they get sample data.

### ScikitLearn

In [ ]:
from sklearn import linear_model
X = [[0., 0.], [1., 1.], [2., 2.], [3., 3.]]
Y = [0., 1., 2., 3.]
reg = linear_model.BayesianRidge()
reg.fit(X, Y)
BayesianRidge()

BayesianRidge(alpha_1=1e-06, alpha_2=1e-06, alpha_init=None,
              compute_score=False, copy_X=True, fit_intercept=True,
              lambda_1=1e-06, lambda_2=1e-06, lambda_init=None, n_iter=300,
              normalize=False, tol=0.001, verbose=False)

In [ ]:
reg.predict([[1, 0.]])


array([0.50000013])

In [ ]:
reg.coef_

array([0.49999993, 0.49999993])

Due to the Bayesian framework, the weights found are slightly different to the ones found by Ordinary Least Squares. However, Bayesian Ridge Regression is more robust to ill-posed problems.

https://scikit-learn.org/stable/modules/linear_model.html#bayesian-ridge-regression


### Package PyMC3 (Probabilistic Programming in Python)

This is a package for Bayesian statistical modeling and probabilistic machine learning

An example of Bayesian linear regression with PyMC 3:
https://docs.pymc.io/notebooks/GLM-linear.html


### Philosophy of Science

Mathematics and Probability theory use deductive inference, i.e., from general statements to particular ones. If the axioms or premises are valid, then the conclusions are certain.

Statistics, Physics, and other empirical sciences use inductive inference, i.e., from particular statements on observations (e.g., sample data) to general statements (e.g., on populations etc). But this is a leap of faith rather than logically justifiable. We are uncertain about conclusions in physics and statistics. This is called Hume's induction problem, after the 18th century Scottish philosopher David Hume ( I am a fan of him! )




<img src="https://media-exp1.licdn.com/dms/image/C4D12AQGB-CLRbvhf4A/article-cover_image-shrink_600_2000/0?e=1610582400&v=beta&t=nArvDx17hLr3H201vg1kd_8GF-AYXR1JN9u5ndabvrI">

The late Bertrand Russell explained it in an entertaining way, especially since **Thanksgiving** is coming soon:

This turkey found that, on his first morning at the turkey farm, he was fed at 9 a.m. However, being a good inductivist, he did not jump to conclusions. He waited until he had collected a large number of observations of the fact that he was fed at 9 a.m., and he made these observations under a wide variety of circumstances, on Wednesdays and Thursdays, on warm days and cold days, on rainy days and dry days. Each day, he added another observation statement to his list. Finally, his inductivist conscience was satisfied and he carried out an inductive inference to conclude, “I am always fed at 9 a.m.”. Alas, this conclusion was shown to be false in no uncertain manner when, on Christmas eve, instead of being fed, he had his throat cut. An inductive inference with true premises has led to a false conclusion. (Alan Chalmers, What is this thing called Science, 2nd edition, University of Queensland Press, St. Lucia, 1982)
